### PR TITLE
Grid based Elastic Distortions (refs: albumentations-team/albumentati…

### PR DESCRIPTION
**Pull request by 4pygmalion**
This pull request introduces the `GridElasticTransformation `for image augmentation, which performs randomized, elastic distortions on images. It includes a comprehensive implementation and documentation for the new transformation.

**New Features:**
`GridElasticTransformation `:
Introduced the GridElasticTransformation for image augmentation, which performs randomized, elastic distortions on images grid by grid.
Supports the application of distortions to both images and masks, ensuring consistent transformations.
image

## Summary by Sourcery

Introduce a new `GridElasticTransformation` for image augmentation, enabling grid-based elastic distortions on images and masks. This includes detailed documentation for the new transformation.

New Features:
- Introduce `GridElasticTransformation` for image augmentation, which performs randomized, elastic distortions on images grid by grid.

Documentation:
- Add comprehensive documentation for the new `GridElasticTransformation` including parameters, targets, and image types.

### CODE REVIEW

File: albumentations/augmentations/geometric/transforms.py [hunk 1]
```diff
@@ -2038,3 +2039,146 @@ def get_params(self) -> dict[str, D4Type]:
 
     def get_transform_init_args_names(self) -> tuple[()]:
         return ()
+
+class GridElasticDeform(DualTransform):
```
Comment (ternaus) [inline albumentations/augmentations/geometric/transforms.py:2043]: You need to add InitPartams class to the tranform that checks input the validity of parameters.

Magintude is non negative,
width and height are > 1

File: albumentations/augmentations/geometric/transforms.py [hunk 2]
```diff
@@ -2038,3 +2039,146 @@ def get_params(self) -> dict[str, D4Type]:
...
+    Targets:
+        image, mask
+
+    Image types:
+        uint8
+    """
+
+    def __init__(
+        self,
+        n_grid_width: int,
```
Comment (ternaus) [inline albumentations/augmentations/geometric/transforms.py:2067]: `n_grid_width` and `n_grid_height`, should be merged into one parameter `num_grid_size` or `num_grid` or `num_grid_xy` or similar.

File: albumentations/augmentations/geometric/transforms.py [hunk 3]
```diff
@@ -2038,3 +2039,146 @@ def get_params(self) -> dict[str, D4Type]:
...
+                    width_of_last_square
+                    if horizontal_tile == self.n_grid_height - 1
+                    else width_of_square
+                )
+                y2 = y1 + (
+                    height_of_last_square
+                    if vertical_tile == self.n_grid_width - 1
+                    else height_of_square
+                )
+                dimensions.append([x1, y1, x2, y2])
```
Comment (ternaus) [inline albumentations/augmentations/geometric/transforms.py:2100]: Better to use

`x_min, y_min, x_max, y_max`, and not `x1, y1, x2, y2`

File: albumentations/augmentations/geometric/transforms.py [hunk 4]
```diff
@@ -2038,3 +2039,146 @@ def get_params(self) -> dict[str, D4Type]:
...
+        )
+
+        polygon_indices = []
+        for i in range((vertical_tiles * horizontal_tiles) - 1):
+            if i not in last_row and i not in last_column:
+                polygon_indices.append(
+                    [i, i + 1, i + horizontal_tiles, i + 1 + horizontal_tiles]
+                )
+
+        for a, b, c, d in polygon_indices:
```
Comment (ternaus) [inline albumentations/augmentations/geometric/transforms.py:2124]: Let's use better naming and not `a`, `b`, `c`, `d`

File: albumentations/augmentations/geometric/transforms.py [hunk 5]
```diff
@@ -2038,3 +2039,146 @@ def get_params(self) -> dict[str, D4Type]:
...
+        magnitude: int,
+        p: float = 1.0,
+        always_apply: bool | None = None,
+    ):
+        super().__init__(p=p, always_apply=always_apply)
+        self.n_grid_width = n_grid_width
+        self.n_grid_height = n_grid_height
+        self.magnitude = abs(magnitude)
+
+    def calculate_dimensions(
```
Comment (ternaus) [inline albumentations/augmentations/geometric/transforms.py:2078]: types for parameters and result

File: albumentations/augmentations/geometric/transforms.py [hunk 6]
```diff
@@ -2038,3 +2039,146 @@ def get_params(self) -> dict[str, D4Type]:
...
+                y2 = y1 + (
+                    height_of_last_square
+                    if vertical_tile == self.n_grid_width - 1
+                    else height_of_square
+                )
+                dimensions.append([x1, y1, x2, y2])
+
+        return dimensions
+
+    def calculate_polygons(self, dimensions, horizontal_tiles, vertical_tiles):
```
Comment (ternaus) [inline albumentations/augmentations/geometric/transforms.py:2104]: types for parameters and result

File: albumentations/augmentations/geometric/transforms.py [hunk 7]
```diff
@@ -2038,3 +2039,146 @@ def get_params(self) -> dict[str, D4Type]:
...
+            polygons[b][2] += dx
+            polygons[b][3] += dy
+            polygons[c][6] += dx
+            polygons[c][7] += dy
+            polygons[d][0] += dx
+            polygons[d][1] += dy
+
+        return polygons
+
+    def generate_mesh(self, polygons, dimensions):
```
Comment (ternaus) [inline albumentations/augmentations/geometric/transforms.py:2139]: types for parameters and result

File: albumentations/augmentations/geometric/transforms.py [hunk 8]
```diff
@@ -2038,3 +2039,146 @@ def get_params(self) -> dict[str, D4Type]:
...
+        return [[dimensions[i], polygons[i]] for i in range(len(dimensions))]
+
+    def distort_image(self, image: np.ndarray, generated_mesh: List[List]):
+        image = Image.fromarray(image)
```
Comment (ternaus) [inline albumentations/augmentations/geometric/transforms.py:2143]: We avoid brigning heavy dependencies to the library when existing could work for the task.

=>

`Pillow` => `opencv`

File: albumentations/augmentations/geometric/transforms.py [hunk 9]
```diff
@@ -2038,3 +2039,146 @@ def get_params(self) -> dict[str, D4Type]:
...
+                image.size, Image.MESH, generated_mesh, resample=Image.BICUBIC
+            )
+        )
+
+    def get_params_dependent_on_data(
+        self, params: Dict[str, Any], data: dict[str, Any]
+    ) -> Dict[str, Any]:
+
+        img = data["image"]
+        h, w = img.shape[:2]
```
Comment (ternaus) [inline albumentations/augmentations/geometric/transforms.py:2155]: not `h`, `w`, but `height`, `width`

**Comment by ternaus** [comment]
Love the transform, but a bit of extra work is needed.

1. Add types for all parameters
2. Add to tests.
3. Add InitSchema class to the transform that verifies the validity of input parameters
4. Remove dependency on Pillow
5. Add to readme.

Check all the code with `pre-commit hook`- here are details on how to work with it: https://github.com/albumentations-team/albumentations/blob/main/CONTRIBUTING.md

**Review by ternaus — CHANGES_REQUESTED** [review]

**Comment by 4pygmalion** [comment]
I'll work on this as a sub-branch of the 4pygmalion
/GridElasticDeformation branch. I'll let you know once it's complete. Thank you for the review.

**Review by ternaus — APPROVED** [review]

**Comment by ternaus** [comment]
@4pygmalion Thank you for the pull request and new Transform

**Pull request merged**